Option field causes C error when struct embeds multiple other structs

### Describe the bug

When struct A embeds struct B and struct C (which has an option field) and struct A is cast to interface A, then comparing the interface A with another interface A will cause a C error.

### Reproduction Steps

```v
fn main() {
	mut arr := []&NodeInterface{}
	arr << &HTMLElement{}
	arr << &HTMLElement{}

	// removing this comparison lets the code compile
	println(arr[0] == arr[1])
}

pub interface Element {
	NodeInterface
mut:
	tag_name string
}

pub interface NodeInterface {
mut:
	node_name string
}

pub struct Node {
pub mut:
	node_name string
}

pub struct HTMLElement {
	Node
pub mut:
	tag_name string
}

pub struct HTMLInputElement {
	// removing either one of these embedded structs lets the code compile
	HTMLElement
	PopoverInvokerElement
pub mut:
	value string
}

pub struct PopoverInvokerElement {
pub mut:
	// changing this field to `&Element` lets the code compile
	popover_target_element ?&Element
	popover_target_action  string
}
```

### Expected Behavior

Expected code to compile and print `true` or `false`

### Current Behavior

```
C:/Users/imado/AppData/Local/Temp/v_0/test.10654267466622163022.tmp.c:510: warning: WINVER redefined
C:/Users/imado/AppData/Local/Temp/v_0/test.10654267466622163022.tmp.c:2319: error: expected struct or union but not 'struct main__Element *'
builder error: 
==================
C error. This should never happen.
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V full version: V 0.4.1 68cbf27.20bce37

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.1 68cbf27.20bce37
OS: windows, Microsoft Windows 11 Pro v22621 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: C:\Users\imado\Documents\cyberian_tiger
vexe: C:\Users\imado\v\v.exe
vexe mtime: 2023-09-25 15:32:08

vroot: OK, value: C:\Users\imado\v
VMODULES: OK, value: C:\Users\imado\.vmodules
VTMP: OK, value: C:\Users\imado\AppData\Local\Temp\v_0

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2023.38-38-g20bce37d
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!IMPORTANT]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority
> for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -130,7 +130,7 @@ fn (mut g Gen) read_field(struct_type ast.Type, field_name string, var_name stri
 @[inline]
 fn (mut g Gen) read_opt_field(struct_type ast.Type, field_name string, var_name string, field_typ ast.Type) string {
 	return if field_typ.has_flag(.option) {
-		field_typ_ := if g.table.sym(field_typ).kind == .interface_ && field_typ.is_ptr() {
+		field_typ_ := if g.table.sym(field_typ).kind in [.interface_, .string] && field_typ.is_ptr() {
 			field_typ.deref()
 		} else {
 			field_typ
@@ -182,14 +182,16 @@ fn (mut g Gen) gen_struct_equality_fn(left_type ast.Type) string {
 			field_type := g.unwrap(field.typ)
 			field_name := c_name(field.name)
 
-			left_arg := g.read_field(left_no_ptr, field_name, 'a')
-			right_arg := g.read_field(left_no_ptr, field_name, 'b')
+			left_arg := g.read_field(left_type, field_name, 'a')
+			right_arg := g.read_field(left_type, field_name, 'b')
 
 			if field_type.sym.kind == .string {
 				if field.typ.has_flag(.option) {
 					left_arg_opt := g.read_opt_field(left_type, field_name, 'a', field.typ)
 					right_arg_opt := g.read_opt_field(left_type, field_name, 'b', field.typ)
 					fn_builder.write_string('(((${left_arg_opt}).len == (${right_arg_opt}).len && (${left_arg_opt}).len == 0) || string__eq(${left_arg_opt}, ${right_arg_opt}))')
+				} else if field.typ.is_ptr() {
+					fn_builder.write_string('((${left_arg}->len == ${right_arg}->len && ${left_arg}->len == 0) || string__eq(*(${left_arg}), *(${right_arg})))')
 				} else {
 					fn_builder.write_string('((${left_arg}.len == ${right_arg}.len && ${left_arg}.len == 0) || string__eq(${left_arg}, ${right_arg}))')
 				}
@@ -554,8 +556,8 @@ fn (mut g Gen) gen_interface_equality_fn(left_type ast.Type) string {
 		g.auto_fn_definitions << fn_builder.str()
 	}
 
-	left_arg := g.read_field(left_no_ptr, '_typ', 'a')
-	right_arg := g.read_field(left_no_ptr, '_typ', 'b')
+	left_arg := g.read_field(left_type, '_typ', 'a')
+	right_arg := g.read_field(left_type, '_typ', 'b')
 
 	fn_builder.writeln('static int v_typeof_interface_idx_${idx_fn}(int sidx); // for auto eq method')
 	fn_builder.writeln('static bool ${fn_name}_interface_eq(${ptr_styp} a, ${ptr_styp} b) {')

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -130,7 +130,17 @@ fn (mut g Gen) read_field(struct_type ast.Type, field_name string, var_name stri
 @[inline]
 fn (mut g Gen) read_opt_field(struct_type ast.Type, field_name string, var_name string, field_typ ast.Type) string {
 	return if field_typ.has_flag(.option) {
-		'*(${g.base_type(field_typ)}*)${g.read_field(struct_type, field_name, var_name)}.data'
+		field_typ_ := if g.table.sym(field_typ).kind == .interface_ && field_typ.is_ptr() {
+			field_typ.deref()
+		} else {
+			field_typ
+		}
+		opt := if g.table.sym(field_typ).kind == .interface_ && field_typ.is_ptr() {
+			'_option_'
+		} else {
+			''
+		}
+		'*(${opt}${g.base_type(field_typ_)}*)${g.read_field(struct_type, field_name, var_name)}.data'
 	} else {
 		g.read_field(struct_type, field_name, var_name)
 	}
@@ -172,8 +182,8 @@ fn (mut g Gen) gen_struct_equality_fn(left_type ast.Type) string {
 			field_type := g.unwrap(field.typ)
 			field_name := c_name(field.name)
 
-			left_arg := g.read_field(left_type, field_name, 'a')
-			right_arg := g.read_field(left_type, field_name, 'b')
+			left_arg := g.read_field(left_no_ptr, field_name, 'a')
+			right_arg := g.read_field(left_no_ptr, field_name, 'b')
 
 			if field_type.sym.kind == .string {
 				if field.typ.has_flag(.option) {
@@ -203,10 +213,38 @@ fn (mut g Gen) gen_struct_equality_fn(left_type ast.Type) string {
 				fn_builder.write_string('${eq_fn}_alias_eq(${left_arg}, ${right_arg})')
 			} else if field_type.sym.kind == .function && !field.typ.has_flag(.option) {
 				fn_builder.write_string('*((voidptr*)(${left_arg})) == *((voidptr*)(${right_arg}))')
-			} else if field_type.sym.kind == .interface_ && !field.typ.has_flag(.option) {
-				ptr := if field.typ.is_ptr() { '*'.repeat(field.typ.nr_muls()) } else { '' }
-				eq_fn := g.gen_interface_equality_fn(field.typ)
-				fn_builder.write_string('${eq_fn}_interface_eq(${ptr}${left_arg}, ${ptr}${right_arg})')
+			} else if field_type.sym.kind == .interface_ {
+				if field.typ.has_flag(.option) {
+					field_typ_ := if field.typ.is_ptr() {
+						field.typ.deref()
+					} else {
+						field.typ
+					}
+					if field.typ.is_ptr() {
+						left_arg_opt := g.read_opt_field(left_type, field_name, 'a', field.typ)
+						right_arg_opt := g.read_opt_field(left_type, field_name, 'b',
+							field.typ)
+						ptr := if field_typ_.is_ptr() {
+							'*'.repeat(field_typ_.nr_muls())
+						} else {
+							''
+						}
+						eq_fn := g.gen_interface_equality_fn(field_typ_)
+						fn_builder.write_string('${eq_fn}_interface_eq(${ptr}${left_arg_opt}, ${ptr}${right_arg_opt})')
+					} else {
+						ptr := if field_typ_.is_ptr() {
+							'*'.repeat(field_typ_.nr_muls())
+						} else {
+							''
+						}
+						eq_fn := g.gen_interface_equality_fn(field_typ_)
+						fn_builder.write_string('${eq_fn}_interface_eq(${ptr}${left_arg}, ${ptr}${right_arg})')
+					}
+				} else {
+					ptr := if field.typ.is_ptr() { '*'.repeat(field.typ.nr_muls()) } else { '' }
+					eq_fn := g.gen_interface_equality_fn(field.typ)
+					fn_builder.write_string('${eq_fn}_interface_eq(${ptr}${left_arg}, ${ptr}${right_arg})')
+				}
 			} else if field.typ.has_flag(.option) {
 				fn_builder.write_string('${left_arg}.state == ${right_arg}.state && !memcmp(&${left_arg}.data, &${right_arg}.data, sizeof(${g.base_type(field.typ)}))')
 			} else {
@@ -516,8 +554,8 @@ fn (mut g Gen) gen_interface_equality_fn(left_type ast.Type) string {
 		g.auto_fn_definitions << fn_builder.str()
 	}
 
-	left_arg := g.read_field(left_type, '_typ', 'a')
-	right_arg := g.read_field(left_type, '_typ', 'b')
+	left_arg := g.read_field(left_no_ptr, '_typ', 'a')
+	right_arg := g.read_field(left_no_ptr, '_typ', 'b')
 
 	fn_builder.writeln('static int v_typeof_interface_idx_${idx_fn}(int sidx); // for auto eq method')
 	fn_builder.writeln('static bool ${fn_name}_interface_eq(${ptr_styp} a, ${ptr_styp} b) {')
@@ -527,7 +565,7 @@ fn (mut g Gen) gen_interface_equality_fn(left_type ast.Type) string {
 		for typ in info.types {
 			fn_builder.writeln('\t\tif (idx == ${typ.idx()}) {')
 			fn_builder.write_string('\t\t\treturn ')
-			match g.table.type_kind(typ) {
+			match g.table.type_kind(typ.set_nr_muls(0)) {
 				.struct_ {
 					eq_fn := g.gen_struct_equality_fn(typ)
 					l_eqfn := g.read_field(left_type, '_${eq_fn}', 'a')

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -53,10 +53,11 @@ fn (mut g Gen) gen_sumtype_equality_fn(left_type ast.Type) string {
 	left := g.unwrap(left_type)
 	ptr_styp := g.typ(left.typ.set_nr_muls(0))
 
-	if left_type in g.generated_eq_fns {
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return ptr_styp
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
 
 	info := left.sym.sumtype_info()
 	g.definitions.writeln('static bool ${ptr_styp}_sumtype_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
@@ -139,10 +140,13 @@ fn (mut g Gen) gen_struct_equality_fn(left_type ast.Type) string {
 	left := g.unwrap(left_type)
 	ptr_styp := g.typ(left.typ.set_nr_muls(0))
 	fn_name := ptr_styp.replace('struct ', '')
-	if left_type in g.generated_eq_fns {
+
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return fn_name
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
+
 	info := left.sym.struct_info()
 	g.definitions.writeln('static bool ${fn_name}_struct_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
 
@@ -199,7 +203,7 @@ fn (mut g Gen) gen_struct_equality_fn(left_type ast.Type) string {
 				fn_builder.write_string('${eq_fn}_alias_eq(${left_arg}, ${right_arg})')
 			} else if field_type.sym.kind == .function && !field.typ.has_flag(.option) {
 				fn_builder.write_string('*((voidptr*)(${left_arg})) == *((voidptr*)(${right_arg}))')
-			} else if field_type.sym.kind == .interface_ {
+			} else if field_type.sym.kind == .interface_ && !field.typ.has_flag(.option) {
 				ptr := if field.typ.is_ptr() { '*'.repeat(field.typ.nr_muls()) } else { '' }
 				eq_fn := g.gen_interface_equality_fn(field.typ)
 				fn_builder.write_string('${eq_fn}_interface_eq(${ptr}${left_arg}, ${ptr}${right_arg})')
@@ -220,10 +224,13 @@ fn (mut g Gen) gen_struct_equality_fn(left_type ast.Type) string {
 fn (mut g Gen) gen_alias_equality_fn(left_type ast.Type) string {
 	left := g.unwrap(left_type)
 	ptr_styp := g.typ(left.typ.set_nr_muls(0))
-	if left_type in g.generated_eq_fns {
+
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return ptr_styp
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
+
 	info := left.sym.info as ast.Alias
 	g.definitions.writeln('static bool ${ptr_styp}_alias_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
 
@@ -277,10 +284,13 @@ fn (mut g Gen) gen_alias_equality_fn(left_type ast.Type) string {
 fn (mut g Gen) gen_array_equality_fn(left_type ast.Type) string {
 	left := g.unwrap(left_type)
 	ptr_styp := g.typ(left.typ.set_nr_muls(0))
-	if left_type in g.generated_eq_fns {
+
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return ptr_styp
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
+
 	elem := g.unwrap(left.sym.array_info().elem_type)
 	ptr_elem_styp := g.typ(elem.typ)
 	g.definitions.writeln('static bool ${ptr_styp}_arr_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
@@ -346,10 +356,13 @@ fn (mut g Gen) gen_array_equality_fn(left_type ast.Type) string {
 fn (mut g Gen) gen_fixed_array_equality_fn(left_type ast.Type) string {
 	left_typ := g.unwrap(left_type)
 	ptr_styp := g.typ(left_typ.typ.set_nr_muls(0))
-	if left_type in g.generated_eq_fns {
+
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return ptr_styp
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
+
 	elem_info := left_typ.sym.array_fixed_info()
 	elem := g.unwrap(elem_info.elem_type)
 	size := elem_info.size
@@ -402,10 +415,13 @@ fn (mut g Gen) gen_fixed_array_equality_fn(left_type ast.Type) string {
 fn (mut g Gen) gen_map_equality_fn(left_type ast.Type) string {
 	left := g.unwrap(left_type)
 	ptr_styp := g.typ(left.typ.set_nr_muls(0))
-	if left_type in g.generated_eq_fns {
+
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return ptr_styp
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
+
 	value := g.unwrap(left.sym.map_info().value_type)
 	ptr_value_styp := g.typ(value.typ)
 	g.definitions.writeln('static bool ${ptr_styp}_map_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
@@ -485,10 +501,13 @@ fn (mut g Gen) gen_interface_equality_fn(left_type ast.Type) string {
 	ptr_styp := g.typ(left.typ.set_nr_muls(0))
 	idx_fn := g.typ(left.typ.set_nr_muls(0).clear_flag(.option))
 	fn_name := ptr_styp.replace('interface ', '')
-	if left_type in g.generated_eq_fns {
+
+	left_no_ptr := left_type.set_nr_muls(0)
+	if left_no_ptr in g.generated_eq_fns {
 		return fn_name
 	}
-	g.generated_eq_fns << left_type
+	g.generated_eq_fns << left_no_ptr
+
 	info := left.sym.info
 	g.definitions.writeln('static bool ${ptr_styp}_interface_eq(${ptr_styp} a, ${ptr_styp} b); // auto')
 

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -972,7 +972,7 @@ pub fn (mut g Gen) write_typeof_functions() {
 			g.writeln('static int v_typeof_interface_idx_${sym.cname}(int sidx) { /* ${sym.name} */ ')
 			for t in inter_info.types {
 				sub_sym := g.table.sym(ast.mktyp(t))
-				g.writeln('\tif (sidx == _${sym.cname}_${sub_sym.cname}_index) return ${int(t)};')
+				g.writeln('\tif (sidx == _${sym.cname}_${sub_sym.cname}_index) return ${int(t.set_nr_muls(0))};')
 			}
 			g.writeln('\treturn ${int(ityp)};')
 			g.writeln('}')

--- a/vlib/v/tests/interface_eq_methods_with_option_and_ref_test.v
+++ b/vlib/v/tests/interface_eq_methods_with_option_and_ref_test.v
@@ -17,5 +17,5 @@ fn test_main() {
 	arr << &Derived{}
 	arr << &Derived{}
 
-	assert arr[0] != arr[1]
+	assert arr[0] == arr[1]
 }

--- a/vlib/v/tests/interface_eq_methods_with_option_and_ref_test.v
+++ b/vlib/v/tests/interface_eq_methods_with_option_and_ref_test.v
@@ -1,0 +1,21 @@
+// for issue 19441
+pub interface Iface {}
+
+pub struct Derived {}
+
+pub struct Struct {
+	field ?&Iface
+}
+
+pub struct Mixin {
+	Derived
+	Struct
+}
+
+fn test_main() {
+	mut arr := []&Iface{}
+	arr << &Derived{}
+	arr << &Derived{}
+
+	assert arr[0] != arr[1]
+}

